pilot/pkg/serviceregistry/aggregate/controller.go

```diff
@@ -167,19 +167,22 @@ func (c *Controller) GetService(hostname host.Name) (*model.Service, error) {
 			out = service.DeepCopy()
 		} else {
 			// If we are seeing the service for the second time, it means it is available in multiple clusters.
-			mergeService(out, service, r.Cluster())
+			mergeService(out, service, r)
 		}
 	}
 	return out, errs
 }
 
-func mergeService(dst, src *model.Service, srcCluster cluster.ID) {
+func mergeService(dst, src *model.Service, srcRegistry serviceregistry.Instance) {
 	dst.Mutex.Lock()
+	defer dst.Mutex.Unlock()
 	if dst.ClusterVIPs == nil {
 		dst.ClusterVIPs = make(map[cluster.ID]string)
 	}
-	dst.ClusterVIPs[srcCluster] = src.Address
-	dst.Mutex.Unlock()
+	// prefer the k8s VIP where possible
+	if srcRegistry.Provider() == provider.Kubernetes || dst.ClusterVIPs[srcRegistry.Cluster()] == "" {
```

Comment:
nit: We have checked provider in L163 https://github.com/istio/istio/pull/34480/files#diff-4ab8bb2188afa1dc914f11b5eaab7e1a544454457d25ef3841fdb6f2c23b1d7bR163

Reply:
Unless it's confusing to leave this unused branch, can we keep it for safety? No strong preference.